File: labs/control-ui/api-shadow/brutal_pacer.ts
export class BrutalPacer {
  public targetBps: number;
  public readonly minBps: number;
  public readonly maxBps: number;
  constructor(targetBps: number, minBps: number = 1000000, maxBps: number = 100000000) {
    this.targetBps = targetBps;
    this.minBps = minBps;
    this.maxBps = maxBps;
  }

  public updateAckFeedback(ackRateBps: number, lossRatio: number): number {
    const boundedLoss = Math.max(0.0, Math.min(1.0, lossRatio));
    // R = AckRate * (1 + loss)
    const compensation = ackRateBps * (1.0 + boundedLoss);
    this.targetBps = Math.max(this.minBps, Math.min(this.maxBps, Math.round(compensation)));
    return this.targetBps;
  }

  public getPacingDelayMs(packetBytes: number): number {
    if (this.targetBps <= 0) return 0;
    return (packetBytes * 8 * 1000) / this.targetBps;
  }
}

export class SalamanderObfuscator {
  private pos = 0;
  private readonly key: Uint8Array;
  constructor(key: Uint8Array) {
    this.key = key;
  }

  public applyInPlace(data: Uint8Array): void {
    const kLen = this.key.length;
    for (let i = 0; i < data.length; i++) {
      const k = this.key[this.pos % kLen] ?? 0;
      data[i] = (data[i] ?? 0) ^ k;
      this.pos++;
    }
  }
}

File: labs/mobile/donor-archive/tunnel/BrutalPacer.kt
// Copyright 2024-2026 LumiNet Authors. All Rights Reserved.
package com.luminet.android.tunnel

class BrutalPacer(
    var targetBps: Long,
    val minBps: Long = 1_000_000,
    val maxBps: Long = 100_000_000
) {
    private val lock = Any()

    fun updateAckFeedback(ackRateBps: Long, lossRatio: Double): Long = synchronized(lock) {
        val boundedLoss = lossRatio.coerceIn(0.0, 1.0)
        // R = AckRate * (1 + loss)
        val compensation = (ackRateBps.toDouble() * (1.0 + boundedLoss)).toLong()
        targetBps = compensation.coerceIn(minBps, maxBps)
        targetBps
    }

    fun getPacingDelayNanos(packetBytes: Int): Long = synchronized(lock) {
        if (targetBps <= 0) return 0L
        ((packetBytes.toLong() * 8.0 * 1_000_000_000L) / targetBps).toLong()
    }
}

class SalamanderObfuscator(private val key: ByteArray) {
    private var pos = 0

    fun applyInPlace(data: ByteArray) {
        for (i in data.indices) {
            data[i] = (data[i].toInt() xor key[pos % key.size].toInt()).toByte()
            pos++
        }
    }
}
